pkg/engine/overlay.go

```diff
@@ -29,11 +33,25 @@ func processOverlay(rule kyverno.Rule, resource unstructured.Unstructured) (resp
 	}()
 
 	patches, err := processOverlayPatches(resource.UnstructuredContent(), rule.Mutation.Overlay)
-	// resource does not satisfy the overlay pattern, we dont apply this rule
-	if err != nil && strings.Contains(err.Error(), "Conditions are not met") {
-		glog.Errorf("Resource %s/%s/%s does not meet the conditions in the rule %s with overlay pattern %s", resource.GetKind(), resource.GetNamespace(), resource.GetName(), rule.Name, rule.Mutation.Overlay)
-		//TODO: send zero response and not consider this as applied?
-		return RuleResponse{}, resource
+	// resource does not satisfy the overlay pattern, we don't apply this rule
+	if err != nil {
+		// condition key is not present in the resource, don't apply this rule
+		// consider as success
+		if strings.Contains(err.Error(), conditionalFieldEmpty) {
```

Comment:
Same as above, these scenarios be identified by handling errors instead of string matching on the message ?
All other engine methods use error value to differentiate different scenarios of error, be consistent. 
If the return structure requires some additional work, can we create custom error types and use helper functions to identify the errors instead?

Reply:
Agreed. 

@realshuting lets add some more return / status codes instead of a string check.

Reply:
The reason I matched by string was this error contains the key that is not present, which I think is helpful to understand the failure:
https://github.com/nirmata/kyverno/blob/1173e062c9a1d13a924d394fceffdba3b4608ef6/pkg/engine/overlayCondition.go#L99
If we don't want to match by the string message, we can return an extra value, i.e. status code to indicate that. In general, we have 2 results:
- success
- failure

And the above can be divided into the following status:
- [success]condition passes: apply overlay
- [success]condition field is not present in the resource: consider as success, message shows in error
- [failure]condition fails: return error
- [success]patch applied successfully
- [failure]fail to apply patch

Reply:
Just to confirm my understanding, in the mutation for overlay we have phases:
- meetConditions
  - [success] condition passes
  - [failure] condition fails 
    - anchorKey is present with value is incorrect
    - some resource(non-anchor) field is not present( report error ?)
  - [failure] anchor key is not present(skip the rule )

- mutateResourceWithOverlay ( apply the overlay on the resource)

@realshuting 
The suggestion was to use not to use string matching to differentiate between the 3 scenarios(success, fail, skip). But instead, use error codes or enums to differentiate the 3 cases. 
- A type with enum values(success, fail, skip) can be used 
- Create 2 custom error types (https://yourbasic.org/golang/create-error/) to differentiate the skip and error scenario and handle them in switch case conditions.

Reply:
>some resource(non-anchor) field is not present( report error ?)

i.e. in the policy you have `(emptyDir): {}` but the volume emptyDir is not present in the resource, this will be considered success and skip process overlay for this rule,  continue applying the next rule. Refer to sample [MutateSafeToEvict](https://github.com/nirmata/kyverno/blob/c63a2b75dbeb775c532194a3911bfb8ce2c42714/samples/best_practices/add_safe-to-evict_annotation.yaml).

For the solution, I'm considering the second option, as it can be used both for meetConditions and processOverlayPatches:

````code
type overlayErr struct {
	statusCode    int
	err           error
}
````

Reply:
the struct contains statusCode and error, and if we are using status code to differentiate between (success, fail, skip) then the error will be empty for success case.

I prefer not adding a new struct, but custom errors. As this is used in some methods in mutation adding a new type struct might increasing complexity. The response from the engine(Validate, Mutate, Generate) uses EngineResponse struct, but internally we use errors to control the flow. But this is more from the code maintainability point of view.

- [success] error== nil
- [failure] check error type
```
switch e := err.(type){
  case *AnchorNotPresent:
  case *ConditionFailed:
  default:
// any other error
}
```

Reply:
>the struct contains statusCode and error, and if we are using status code to differentiate between (success, fail, skip) then the error will be empty for success case.

No, as still, it is helpful to show the detailed message why we skip processing this rule:
https://github.com/nirmata/kyverno/blob/c63a2b75dbeb775c532194a3911bfb8ce2c42714/pkg/engine/overlay.go#L40-L42

In either solution we will add new struct `AnchorNotPresent `, `ConditionFailed ` or `overlayErr`, change error message from type `error` to `string` may be more accurate.
````code

const (
	conditionFailure codeKey = iota
	conditionNotPresent
	conditionSuccess
	overlayFailure
)

type overlayErr struct {
	statusCode    codeKey
	msg           string
}
````

Reply:
ok.
I meant the error type is just to identify the error type, we still have the error message. Which is what we currently have, I just suggested replacing string comparison with error type check the rest will remain the same. Less code to modify and test :)

But introducing a new struct is fine too.